### PR TITLE
Feature/5701 is based on

### PR DESCRIPTION
**Pull request by saminahbab**
### What

[Trello Ticket Here](https://trello.com/c/o7gox9jw/5701-add-check-to-area-types-endpoint-to-ensure-population-type-is-published-for-public-users)

- Ammended the GET population-types/<pop-type>/area-types endpoint. 
- Added private endpoints, and authorization middleware. 
- Changed logic to return only published datasets to unauthorised users. 
- Added logic that makes public request to dataset-api to determine if published datasets exist for given population types.
- Ammended Unit Tests 
- Added functionality to Component Tests to represent new server setup.
- Added Component Tests to test above 

### How to review

run `make test` and `make component-test`

### Who can review

Team B

### CODE REVIEW

File: handler/population-types.go [hunk 1]
```diff
@@ -35,27 +43,68 @@ func (h *PopulationTypes) Get(w http.ResponseWriter, req *http.Request) {
...
 		return
 	}
 
 	var resp contract.GetAreaTypesResponse
 
+	if !authenticated {
+		datasets, err := h.datasetClient.GetDatasets(ctx,
+			"",
+			"",
+			// What do I put here for collection ID ?!
```
Comment (saminahbab) [inline handler/population-types.go:83]: Here, I was not sure where I get the collection ID from, could someone please provide advice?

File: handler/population-types_test.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ func TestEndpointRoot(t *testing.T) {
 		cantabularClient := fakeCantabularClient{
 			listDatasetsReturnData: []string{"Thing one", "Thing two"},
 		}
-		handler := handler.NewPopulationTypes(responder, &cantabularClient)
+		handler := handler.NewPopulationTypes(responder, &cantabularClient, nil, nil)
```
Comment (PauloPortugal) [inline handler/population-types_test.go:27]: Should we be testing the behavior of the other clients?

**Comment by saminahbab** [comment]
I have added a test to show a failing dataset client, but I feel as if they
are covered by unit tests in their respective unit tests in api-clients-go.

File: handler/population-types_test.go (hunk 2 above)
Comment (saminahbab) [reply handler/population-types_test.go:27]: I have added a test to show a failing dataset client, but I feel as if they
are covered by unit tests in their respective unit tests in api-clients-go.

File: handler/population-types.go (hunk 1 above)
Comment (PauloPortugal) [reply handler/population-types.go:83]: I don't think you need any `collectionID` here, as it is going to be added to the Header.
We may need the auth and service token though

Comment (saminahbab) [reply handler/population-types.go:83]: no need for auth and service, as it will be an unauthenticated request

Comment (saminahbab) [reply handler/population-types.go]: Will then leave this to be empty string and resolve, cheers!

File: middleware/auth.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
+package middleware
```
[the remark this answers is not in the document]
Comment (saminahbab) [reply middleware/auth.go:1]: But is it not already implicitly tested through component tests? There are now steps that differentiate between public and private endpoints and expected results conditioned on them

File: config/config.go [hunk 4]
```diff
@@ -12,12 +12,18 @@ type Config struct {
 	GracefulShutdownTimeout    time.Duration `envconfig:"GRACEFUL_SHUTDOWN_TIMEOUT"`
 	HealthCheckInterval        time.Duration `envconfig:"HEALTHCHECK_INTERVAL"`
 	HealthCheckCriticalTimeout time.Duration `envconfig:"HEALTHCHECK_CRITICAL_TIMEOUT"`
+	EnablePrivateEndpoints     bool          `envconfig:"ENABLE_PRIVATE_ENDPOINTS"`
+
```
Comment (ChrysmOre) [inline config/config.go:16]: No need for this blank line

File: config/config.go [hunk 5]
```diff
@@ -36,6 +42,12 @@ func Get() (*Config, error) {
 		GracefulShutdownTimeout:    5 * time.Second,
 		HealthCheckInterval:        30 * time.Second,
 		HealthCheckCriticalTimeout: 90 * time.Second,
+		EnablePrivateEndpoints:     true,
+		ZebedeeURL:                 "http://localhost:8082",
+		ServiceAuthToken:           "",
+		EnablePermissionsAuth:      false,
+		// NOTE: ensure that this is routed to the *unauthenticated* dataset endpoint.
+		DatasetAPIURL: "http://localhost:22000",
```
Comment (ChrysmOre) [inline config/config.go:50]: There isn't a separate endpoint that is configured here, it just depends on how the service is set up (either on web or publishing) so this comment can be removed

File: config/config.go [hunk 6]
```diff
@@ -36,6 +42,12 @@ func Get() (*Config, error) {
 		GracefulShutdownTimeout:    5 * time.Second,
 		HealthCheckInterval:        30 * time.Second,
 		HealthCheckCriticalTimeout: 90 * time.Second,
+		EnablePrivateEndpoints:     true,
+		ZebedeeURL:                 "http://localhost:8082",
+		ServiceAuthToken:           "",
+		EnablePermissionsAuth:      false,
```
Comment (ChrysmOre) [inline config/config.go:48]: Probably better to have permissionsAuth default to true?

File: features/area-types-private.feature [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    Given I have some population types in cantabular
+    And a geography query response is available from Cantabular api extension
+
+    And the following response is returned from dp-dataset-api:
+    """
+    {"total_count": 0}
+    """
+
+
```
Comment (ChrysmOre) [inline features/area-types-private.feature:45]: Some extra blank lines here

File: handler/population-types.go [hunk 8]
```diff
@@ -35,27 +43,67 @@ func (h *PopulationTypes) Get(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
-// Get is the handler for GET /area-types
+/*
+   GET population-type/<id>/area-types
+   Gets area types from cantabular.
+   If not authenticated, then only return
```
Comment (ChrysmOre) [inline handler/population-types.go:49]: This comment can probably be removed

File: handler/population-types.go [hunk 9]
```diff
@@ -35,27 +43,67 @@ func (h *PopulationTypes) Get(w http.ResponseWriter, req *http.Request) {
...
+				ctx,
+				w,
+				http.StatusInternalServerError,
+				errors.Wrap(err, "failed to get area-types"),
+			)
+			return
+
+		}
+		if datasets.TotalCount == 0 {
+			// No published datasets
```
Comment (ChrysmOre) [inline handler/population-types.go:97]: Comment seems slightly superfluous

File: handler/population-types.go [hunk 10]
```diff
@@ -68,3 +116,29 @@ func (h *PopulationTypes) GetAreaTypes(w http.ResponseWriter, r *http.Request) {
 
 	h.responder.JSON(ctx, w, http.StatusOK, resp)
 }
+
+// Required for GET area types
+func (h *PopulationTypes) authenticate(r *http.Request, logData log.Data) bool {
```
Comment (ChrysmOre) [inline handler/population-types.go:121]: This logData is not actually being used for anything here, it should be removed

File: service/routes.go [hunk 11]
```diff
@@ -15,14 +18,41 @@ func (svc *Service) buildRoutes(ctx context.Context) {
...
+	checkIdentity := dphandlers.IdentityWithHTTPClient(svc.identityClient)
+
+	r.Use(checkIdentity)
+	r.Use(middleware.LogIdentity())
+	r.Use(permissions.Require(auth.Permissions{Read: true}))
+
+	r.Get("/population-types", populationTypes.Get)
+	r.Get("/population-types/{population-type}/area-types", populationTypes.GetAreaTypes)
+
+	// added from cant filt flex
```
Comment (ChrysmOre) [inline service/routes.go:56]: Not a good comment to leave hanging about, it just says 'I copied this from somewhere else, I don't actually know what it does'

**Review by ChrysmOre — COMMENTED** [review]
There's a comment about needing to use the unauthorized version of the datasets endpoint but I don't think this will be possible if this service is deployed to publishing. A rethink will be needed there if that's the case

File: config/config.go (hunk 6 above)
Comment (saminahbab) [reply config/config.go:48]: I would err on the side of caution and leave as it. I think people are instructed in general to set ENABLE_PERMISSIONS_AUTH=true in their rc files, and this interferes with component tests. I have tried setting `os.clearenv()` in the `main_test.go` file, to no avail, so would keep as is, unless you know a way around that

File: config/config.go (hunk 5 above)
Comment (saminahbab) [reply config/config.go:50]: will remove, but given the slightly roundabout way in which we are trying to get published datasets, I would provide as much info as possible, maybe rephrase the above more in line with your comment

File: handler/population-types.go (hunk 8 above)
Comment (saminahbab) [reply handler/population-types.go:49]: Again, given the roundabout way we are trying to get the unpublished datasets, i think this comment should definitely be there for future reference

File: handler/population-types.go (hunk 9 above)
Comment (saminahbab) [reply handler/population-types.go:97]: agreed, will remove

File: handler/population-types.go (hunk 10 above)
Comment (saminahbab) [reply handler/population-types.go:121]: removed

File: main_test.go [hunk 12]
```diff
@@ -24,13 +25,23 @@ type ComponentTest struct {
...
 		panic(err)
 	}
+	apiFeature := componenttest.NewAPIFeature(component.InitialiseService)
+	component.APIFeature = apiFeature
 
 	ctx.Before(func(ctx context.Context, sc *godog.Scenario) (context.Context, error) {
-		component.Reset()
+		apiFeature.Reset()
+		if err := component.Reset(); err != nil {
+			glog.Panicf("unable to initialise scenario: %s", err)
```
[the remark this answers is not in the document]
Comment (saminahbab) [reply main_test.go:41]: was copying filter-flex-api as my reference point. will return error instead. Though if you cannot set up/reset scenarios without errorring, then a panic seems fine to me.

File: features/area-types.feature [hunk 13]
```diff
@@ -1,14 +1,63 @@
...
     Then an internal server error saying "failed to get area-types: error(s) returned by graphQL query" is returned
     And the HTTP status code should be "404"
+
+  Scenario: Dataset Client returns errors
+    Given the dp-dataset-api is returning errors
+    And a geography query response is available from Cantabular api extension
+
+    When I GET "/population-types/Example/area-types"
+    Then the HTTP status code should be "500"
+    And an internal server error saying "failed to get area-types" is returned
```
[the remark this answers is not in the document]
Comment (saminahbab) [reply features/area-types.feature]: I'll leave the ones I have not touched for your merge to deal with then

**Comment by saminahbab** [comment]
> There's a comment about needing to use the unauthorized version of the datasets endpoint but I don't think this will be possible if this service is deployed to publishing. A rethink will be needed there if that's the case

@ChrysmOre  do we know the person to ask about this? it would be better to know asap so that we can start on other approaches

File: middleware/auth.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrysmOre) [reply middleware/auth.go:1]: The attitude we've been asked to take is 'component tests first, unit tests to fill in the gaps'. I don't have particularly strong feelings either way in this instance. I think maybe a unit test specifically for NewPermissions might be a bit much though.

File: features/area-types.feature (hunk 13 above)
[the remark this answers is not in the document]
Comment (saminahbab) [reply features/area-types.feature]: Will be resolving this if no objections.

File: features/steps/population_types_component.go [hunk 14]
```diff
@@ -27,36 +28,44 @@ type PopulationTypesComponent struct {
...
 	HttpServer                   *http.Server
-	apiFeature                   *componenttest.APIFeature
+	APIFeature                   *componenttest.APIFeature
 	fakeCantabularDatasets       []string
 	fakeCantabularIsUnresponsive bool
 	fakeCantabularGeoDimensions  *cantabular.GetGeographyDimensionsResponse
 	service                      *service.Service
 	InitialiserMock              service.Initialiser
+
+	// Here add identity client?
```
Comment (ChrysmOre) [inline features/steps/population_types_component.go:38]: Nope, comment can be removed

File: features/area-types.feature [hunk 15]
```diff
@@ -1,14 +1,69 @@
...
-    Then a list of area types is returned
-    And the HTTP status code should be "200"
+  Background:
+    Given private endpoints are not enabled
 
+  Scenario: Getting published area types
+    Given the following response is returned from dp-dataset-api:
+    """
+    {
+    "total_count": 1
```
Comment (ChrysmOre) [inline features/area-types.feature:10]: Needs indent

File: features/area-types.feature [hunk 16]
```diff
@@ -1,14 +1,69 @@
...
+          }
+        ]
+    }
+    """
+
+  Scenario: Getting unpublished area types
+    Given the following response is returned from dp-dataset-api:
+    """
+    {
+    "total_count": 0
```
Comment (ChrysmOre) [inline features/area-types.feature:41]: Here too

File: features/steps/population_types_component.go [hunk 17]
```diff
@@ -27,36 +28,44 @@ type PopulationTypesComponent struct {
...
 	return c, nil
 }
 
-func (c *PopulationTypesComponent) Reset() *PopulationTypesComponent {
-	c.apiFeature.Reset()
-	return c
+func (c *PopulationTypesComponent) Reset() error {
+	c.APIFeature.Reset()
+	if _, err := c.InitialiseService(); err != nil {
+		return err
```
Comment (ChrysmOre) [inline features/steps/population_types_component.go:66]: Error could do with being wrapped

File: features/area-types.feature (hunk 15 above)
Comment (saminahbab) [reply features/area-types.feature]: done

**Review by ChrysmOre — APPROVED** [review]
